Nodejs aws lambda with esbuild bundling

mneil opened the pull request.

3p3r commented on .gitignore:
```diff
@@ -254,3 +254,9 @@ aws-cdk-web/
+
+dist/esbuild.wasm
+public/esbuild.wasm
+lib/construct/cdk-web-lambda/go-wasm.js
```
this is ignored here, but also included in this diff? it should not be here I think.
I think the ignore path is incorrect

3p3r commented on cdk-web.webpack.config.js:
```diff
@@ -8,17 +8,15 @@ const $ = (s = "") => path.resolve(common.__ROOT, s);
 module.exports = {
   node: {
-    os: true,
+    crypto: true,
```
what prompted this change?

mneil replied:
reverting

3p3r commented on cdk-web.webpack.config.js:
```diff
@@ -8,17 +8,15 @@ const $ = (s = "") => path.resolve(common.__ROOT, s);
     https: true,
     stream: true,
     console: true,
-    process: "mock",
```
what prompted this change?

mneil replied:
reverting

3p3r commented on cdk-web.webpack.config.js:
```diff
@@ -8,17 +8,15 @@ const $ = (s = "") => path.resolve(common.__ROOT, s);
     tls: "mock",
     net: "mock",
     zlib: true,
-    path: true,
```
what prompted this change?

mneil replied:
reverting

3p3r commented on webpack/modules/fs.js:
```diff
@@ -1,3 +1,4 @@
 const memfs = require("memfs");
 
 module.exports = { ...memfs };
+globalThis.fs = memfs;
```
do you need this anymore? is this used inside go-wasm.js?

mneil replied:
Yes, still need this. The go-wasm.js doesn't require fs, it looks for globalThis.fs and if it is undefined it creates it's own globalThis.fs. I opted to add it here rather than trying to replace the checks for globalThis.fs. I think if I do the other changes you suggested so we don't need the prebuild plugin then this way would be the only way to get the correct fs patch for the go-wasm.js file.

3p3r replied:
since this bursts the bubble that webpack creates around our UMD module, after I write tests I'll try to see if I can somehow get it to work properly now that `go-wasm.js` is actually passing through webpack.

3p3r commented on cdk-web.webpack.config.js:
```diff
@@ -65,13 +63,20 @@ module.exports = {
   },
   plugins: [
     ...(common.__SERVER ? [new plugins.WebpackMildCompile()] : []),
+    plugins.PreBuildPlugin,
```
for the sake of staying consistent please subclass this plugin and `new` it here to stay coherent with what's below it.

mneil replied:
So the prebuild plugin doesn't allow you to create a new instance of it this way without doing all the config here. I was going to subclass it, overload the constructor, pass in defaults, then read them in during the create method... but that felt like way too much work. Anyway, going to remove it based on your other ideas.

3p3r commented on webpack/modules/aws-cdk-lib/aws-lambda-nodejs/lib/esbuild.js:
```diff
@@ -0,0 +1,45 @@
+// Load the go wasm.
+require("./go-wasm");
```
if you are not planning to actually modify this file, instead of copying it in the pre-build plugin, you can simply alias "go-wasm" to the original one inside node_modules with webpack and just call `require("go-wasm")` here.

3p3r commented on webpack/modules/aws-cdk-lib/aws-lambda-nodejs/lib/esbuild.js:
```diff
@@ -0,0 +1,45 @@
+  constructor() {
+    super();
+    this.wasm = undefined;
+    this.workDir = "/app";
```
shouldn't this be `process.env.CDK_OUTDIR` ?

mneil replied:
No, `/app` was correct. It should be `cwd` but it's actually unused here now. Removing it.

3p3r commented on webpack/modules/aws-cdk-lib/aws-lambda-nodejs/lib/esbuild.js:
```diff
@@ -0,0 +1,45 @@
+    this.wasm = undefined;
+    this.workDir = "/app";
+    this.user = "esbuild";
+    this.homeDir = "/home/esbuild";
```
I'd suggest we used the patched "os" module here to get the mock user and home values

mneil replied:
forgot this was here. Not needed for this. I used these with the node version of the go wasm exec for esbuild but that went nowhere. I am removing it.

3p3r commented on webpack/modules/aws-cdk-lib/aws-lambda-nodejs/lib/esbuild.js:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  async load() {
+    const response = await fetch("esbuild.wasm");
```
how does this work? "esbuild.wasm" is externalized here?

mneil replied:
esbuild.wasm is external. The prebuild plugin moves it into dist and public.

3p3r commented on webpack/modules/aws-cdk-lib/core/lib/asset-staging.js:
```diff
@@ -0,0 +1,41 @@
+    fs.mkdirSync(bundleDir, { recursive: true });
+    let localBundling;
+    try {
+      console.log(`Bundling asset ${this.node.path}... in ${bundleDir}\n`);
```
avoid using `console.*` please. `debug` should be used instead. we don't want logs in browser by default: https://github.com/3p3r/cdk-web/blob/main/docs/logs.md

mneil replied:
Like another tool we may or may not have made I think we need a logger . This log is part of cdk core. I will change to debug though. https://github.com/aws/aws-cdk/blob/v1.152.0/packages/@aws-cdk/core/lib/asset-staging.ts#L444 it is written to stderr.

3p3r replied:
wowow I can't believe they're directly writing to stderr. cdk has a logger that I patched. might have been in the cli repo. anyhoo, calls to console.log and process.std** streams are all nuked by webpack FYI.

3p3r commented on webpack/modules/aws-cdk-lib/aws-lambda-nodejs/lib/esbuild.js:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  async load() {
+    const response = await fetch("esbuild.wasm");
```
now that you are using `fetch` here, I have to bring in babel. `fetch` is definitely not supported in browser environments we're trying to support.

mneil replied:
fetch was browser first then made its way into node. It's 100% a browser supported method. https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch . It was meant to simplify XMLHTTPRequest.

Support since Chrome 42: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#browser_compatibility

3p3r replied:
wasn't talking about `fetch` in particular but just in general our internal sources are growing so it's a good idea to use babel-loader for our sources only and then enforce a code style

3p3r commented on webpack/plugins/pre-build-plugin.js:
```diff
@@ -0,0 +1,27 @@
+  build: (compiler, compilation, matchedFiles) => {
+    debug("copying esbuild to dist");
+    const esbuildRoot = path.resolve(require.resolve("esbuild-wasm"), "../..");
+    fs.copyFileSync(path.resolve(esbuildRoot, "esbuild.wasm"), path.resolve(__ROOT, "dist/esbuild.wasm"));
```
up there somewhere I left a comment about this. we can have webpack to do this proper for us.

- since `esbuild.wasm` is not really part of the bundle, it does not need to be copied here. it can be moved to post build plugin.
- `go-wasm.js` is used by sources so we can just alias `go-wasm`: `path.resolve(esbuildRoot, "wasm_exec.js")` and then in sources you can just do `require("go-wasm")`.

if you do the above ^, you will not need a pre build plugin :)

3p3r commented on package.json:
```diff
@@ -65,12 +66,14 @@
     "npm-dts": "^1.3.11",
     "npm-run-all": "^4.1.5",
     "os-browserify": "^0.3.0",
+    "prebuild-webpack-plugin": "^1.1.1",
```
left a comment down. you can remove this package.

3p3r commented:
this right here possibly is one of the most important PRs in the DevOps field for the upcoming years.

BEHOLD WEB CONSTRUCTS! BEHOLD WEB LAMBDAS! 🎉

mneil commented:
I've updated based on your feedback. I removed the prebuild plugin entirely. Mostly just did the things you asked other than:

Fetch is supported, no need for babel
I made the loading of esbuild.wasm a singleton so that you only get 1 binary for n lambdas. Had to do this with prototype so it's a bit ugly looking but it works. Can change to static members if we add babel.
